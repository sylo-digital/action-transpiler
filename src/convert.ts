import { replaceAdvancedEmbed } from "./conversions/replace-advancedembed";
import { replaceArgs } from "./conversions/replace-args";
import { replaceBrackets } from "./conversions/replace-brackets";
import { replaceChoose } from "./conversions/replace-choose";
import { replaceCommandTags } from "./conversions/replace-command-tags";
import { replaceCreatedAt } from "./conversions/replace-createdAt";
import { replaceGeneric } from "./conversions/replace-generic";
import { replaceLowerUpper } from "./conversions/replace-lower-upper";
import { replaceMessageTimestamp } from "./conversions/replace-message-timestamp";
import { replaceNotesWithComments } from "./conversions/replace-notes-with-comments";
import { replaceRandstr } from "./conversions/replace-randstr";
import { replaceVariables } from "./conversions/replace-variables";
import { treeToText } from "./helpers/tree-to-text";
import { Lexer } from "./lexer";
import { Converter, ScriptOptions } from "./types";

const converters: Converter[] = [
  replaceGeneric,
  replaceArgs,
  replaceNotesWithComments,
  replaceBrackets,
  replaceVariables,
  replaceCommandTags,
  replaceMessageTimestamp,
  replaceCreatedAt,
  replaceChoose,
  replaceLowerUpper,
  replaceRandstr,
  replaceAdvancedEmbed, // must be last
];

export function convert(script: string) {
  const options: ScriptOptions = {
    appendAfter: [],
    appendBefore: [],
    source: script,
    warnings: [],
    errors: [],
    injectInputOption: false,
  };

  const lexer = new Lexer(script);
  const tree = lexer.parse();
  for (const converter of converters) {
    converter(tree, options);
  }

  let text = treeToText(tree);
  if (options.warnings[0]) {
    text = `${options.warnings.map((warning) => `// ${warning}`).join("\n")}\n${text}`;
  }

  if (options.appendAfter) {
    text = options.appendBefore.join("\n") + "\n" + text;
  }

  if (options.appendAfter[0]) {
    text = text + "\n" + options.appendAfter.join("\n");
  }

  return {
    output: text.trim(),
    warnings: options.warnings,
    errors: options.errors,
    injectInputOption: options.injectInputOption,
  };
}
