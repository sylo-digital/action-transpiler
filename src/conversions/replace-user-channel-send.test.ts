import { convert } from "../convert";

it("should rename {user.send} and inject the polyfill", () => {
  expect(convert("{user.send}")).toMatchSnapshot();
  expect(convert("{user.send;message}")).toMatchSnapshot();
  expect(convert("{user.send;message;user}")).toMatchSnapshot();
});

it("should rename {channel.send} and inject the polyfill", () => {
  expect(convert("{channel.send}")).toMatchSnapshot();
  expect(convert("{channel.send;message}")).toMatchSnapshot();
  expect(convert("{channel.send;message;channel}")).toMatchSnapshot();
});
