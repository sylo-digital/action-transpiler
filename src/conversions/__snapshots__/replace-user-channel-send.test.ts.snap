// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should rename {channel.send} and inject the polyfill 1`] = `
Object {
  "errors": Array [],
  "injectInputOption": false,
  "output": "// This function has been injected to replace functionality of a legacy {channel.send} tag.
[#function;legacy_channel_send;{=message};{=channel}]
    {responder.text;{$message}} // Add the text to the responder
    {=channelId;{or;{$channel};{channel.id}}} // Get the channel we're sending to
    {responder.channel;{$channelId}} // Configure the responder to send the message to the channel
    {responder.send} // Send the message
[/function]
{legacy_channel_send}",
  "warnings": Array [],
}
`;

exports[`should rename {channel.send} and inject the polyfill 2`] = `
Object {
  "errors": Array [],
  "injectInputOption": false,
  "output": "// This function has been injected to replace functionality of a legacy {channel.send} tag.
[#function;legacy_channel_send;{=message};{=channel}]
    {responder.text;{$message}} // Add the text to the responder
    {=channelId;{or;{$channel};{channel.id}}} // Get the channel we're sending to
    {responder.channel;{$channelId}} // Configure the responder to send the message to the channel
    {responder.send} // Send the message
[/function]
{legacy_channel_send;message}",
  "warnings": Array [],
}
`;

exports[`should rename {channel.send} and inject the polyfill 3`] = `
Object {
  "errors": Array [],
  "injectInputOption": false,
  "output": "// This function has been injected to replace functionality of a legacy {channel.send} tag.
[#function;legacy_channel_send;{=message};{=channel}]
    {responder.text;{$message}} // Add the text to the responder
    {=channelId;{or;{$channel};{channel.id}}} // Get the channel we're sending to
    {responder.channel;{$channelId}} // Configure the responder to send the message to the channel
    {responder.send} // Send the message
[/function]
{legacy_channel_send;message;channel}",
  "warnings": Array [],
}
`;

exports[`should rename {user.send} and inject the polyfill 1`] = `
Object {
  "errors": Array [],
  "injectInputOption": false,
  "output": "// This function has been injected to replace functionality of a legacy {user.send} tag.
[#function;legacy_user_send;{=message};{=user}]
    {responder.text;{$message}} // Add the text to the responder
    {=userId;{or;{$user};{user.id}}} // Get the user we're sending to
    {responder.dm;{$userId}} // Configure the responder to direct-message the user
    {responder.send} // Send the message
[/function]
{legacy_user_send}",
  "warnings": Array [],
}
`;

exports[`should rename {user.send} and inject the polyfill 2`] = `
Object {
  "errors": Array [],
  "injectInputOption": false,
  "output": "// This function has been injected to replace functionality of a legacy {user.send} tag.
[#function;legacy_user_send;{=message};{=user}]
    {responder.text;{$message}} // Add the text to the responder
    {=userId;{or;{$user};{user.id}}} // Get the user we're sending to
    {responder.dm;{$userId}} // Configure the responder to direct-message the user
    {responder.send} // Send the message
[/function]
{legacy_user_send;message}",
  "warnings": Array [],
}
`;

exports[`should rename {user.send} and inject the polyfill 3`] = `
Object {
  "errors": Array [],
  "injectInputOption": false,
  "output": "// This function has been injected to replace functionality of a legacy {user.send} tag.
[#function;legacy_user_send;{=message};{=user}]
    {responder.text;{$message}} // Add the text to the responder
    {=userId;{or;{$user};{user.id}}} // Get the user we're sending to
    {responder.dm;{$userId}} // Configure the responder to direct-message the user
    {responder.send} // Send the message
[/function]
{legacy_user_send;message;user}",
  "warnings": Array [],
}
`;
